compiler/rustc_builtin_macros/src/format.rs

```diff
@@ -744,78 +745,93 @@ impl<'a, 'b> Context<'a, 'b> {
         // format "string"
         let pieces = self.ecx.expr_vec_slice(self.fmtsp, self.str_pieces);
 
-        // Before consuming the expressions, we have to remember spans for
-        // count arguments as they are now generated separate from other
-        // arguments, hence have no access to the `P<ast::Expr>`'s.
-        let spans_pos: Vec<_> = self.args.iter().map(|e| e.span).collect();
-
-        // Right now there is a bug such that for the expression:
-        //      foo(bar(&1))
-        // the lifetime of `1` doesn't outlast the call to `bar`, so it's not
-        // valid for the call to `foo`. To work around this all arguments to the
-        // format! string are shoved into locals. Furthermore, we shove the address
-        // of each variable because we don't want to move out of the arguments
-        // passed to this function.
-        for (i, e) in self.args.into_iter().enumerate() {
-            for arg_ty in self.arg_unique_types[i].iter() {
-                args.push(Context::format_arg(self.ecx, self.macsp, e.span, arg_ty, i));
-            }
-            // use the arg span for `&arg` so that borrowck errors
-            // point to the specific expression passed to the macro
-            // (the span is otherwise unavailable in MIR)
-            heads.push(self.ecx.expr_addr_of(e.span.with_ctxt(self.macsp.ctxt()), e));
-        }
-        for index in self.count_args {
-            let span = spans_pos[index];
-            args.push(Context::format_arg(self.ecx, self.macsp, span, &Count, index));
-        }
-
-        let args_array = self.ecx.expr_vec(self.macsp, args);
-
-        // Constructs an AST equivalent to:
-        //
-        //      match (&arg0, &arg1) {
-        //          (tmp0, tmp1) => args_array
-        //      }
+        // We need to construct a &[ArgumentV1] to pass into the fmt::Arguments
+        // constructor. In general the expressions in this slice might be
+        // permuted from their order in original_args (such as in the case of
+        // "{1} {0}"), or may have multiple entries referring to the same
+        // element of original_args ("{0} {0}").
         //
-        // It was:
+        // The following Iterator<Item = (usize, &ArgumentType)> has one item
+        // per element of our output slice, identifying the index of which
+        // element of original_args it's passing, and that argument's type.
+        let fmt_arg_index_and_ty = self
+            .arg_unique_types
+            .iter()
+            .enumerate()
+            .flat_map(|(i, unique_types)| unique_types.iter().map(move |ty| (i, ty)))
+            .chain(self.count_args.iter().map(|i| (*i, &Count)));
+
+        // Figure out whether there are permuted or repeated elements. If not,
+        // we can generate simpler code.
+        let nicely_ordered = fmt_arg_index_and_ty
+            .clone()
+            .is_sorted_by(|(i, _), (j, _)| (i < j).then_some(Ordering::Less));
```

Comment:
That's very subtle. It would be great if you could add a comment. And maybe find a simpler (if more verbose) way to express this. I personally cannot decode what's going on here in a reasonable amount of time.

Reply:
`array_windows` (https://github.com/rust-lang/rust/issues/75027) may be clearer for this. The sequence has no indices out of order or repeated if: in every adjacent pair of elements, the first one's index is less than the second one's index.

https://github.com/rust-lang/rust/blob/e2ef89707ea7b77843d2364992d77b7f30bddae5/compiler/rustc_builtin_macros/src/format.rs#L772-L775